dataplane-drivers/windataplane/endpoint_mgr.go

```diff
@@ -0,0 +1,309 @@
+	envNetworkName = "KUBE_NETWORK"
+	// the default hns network name to use if the envNetworkName environment
+	// variable does not resolve to a value
+	defaultNetworkName = "l2bridge"
+)
+
+var (
+	ErrorUnknownEndpoint = errors.New("Endpoint could not be found")
+	ErrorUpdateFailed = errors.New("Endpoint update failed")
+)
+
+// endpointManager processes WorkloadEndpoint* updates from the datastore. Updates are
+// stored and pended for processing during CompleteDeferredWork. endpointManager is also
+// responsible for orchestrating a refresh of all impacted endpoints after a IPSet update.
+type endpointManager struct {
+	// the name of the hns network for which we will be managing endpoint policies.
+	hnsNetworkName string
+	// the policysets dataplane to be used when looking up endpoint policies/profiles.
+	policysetsDataplane policysets.PolicySetsDataplane
+	// pendingWlEpUpdates stores any pending updates to be performed per endpoint.
+	pendingWlEpUpdates map[proto.WorkloadEndpointID]*proto.WorkloadEndpoint
+	// activeWlEndpoints stores the active/current state that was applied per endpoint
+	activeWlEndpoints map[proto.WorkloadEndpointID]*proto.WorkloadEndpoint
+	// addressToEndpointId serves as a hns endpoint id cache. It enables us to lookup the hns
+	// endpoint id for a given endpoint ip address.
+	addressToEndpointId map[string]string
+	// read-write lock to protect access to the addressToEndpointId map during cache updates.
+	endpointLock sync.RWMutex
+	// lastCacheUpdate records the last time that the addressToEndpointId map was refreshed.
+	lastCacheUpdate time.Time
+}
+
+func newEndpointManager(policysets policysets.PolicySetsDataplane) *endpointManager {
+	var networkName string
+	if os.Getenv(envNetworkName) != "" {
+		networkName = os.Getenv(envNetworkName)
+		log.WithField("NetworkName", networkName).Info("Setting hns network name from environment variable")
+	} else {
+		networkName = defaultNetworkName
+		log.WithField("NetworkName", networkName).Info("No Network Name environment variable was found, using default name")
+	}
+
+	return &endpointManager{
+		hnsNetworkName: networkName,
+		policysetsDataplane: policysets,
+		addressToEndpointId: make(map[string]string),
+		activeWlEndpoints: map[proto.WorkloadEndpointID]*proto.WorkloadEndpoint{},
+		pendingWlEpUpdates: map[proto.WorkloadEndpointID]*proto.WorkloadEndpoint{},
+	}
+}
+
+// OnUpdate is called by the main dataplane driver loop during the first phase. It processes
+// specific types of updates from the datastore.
+func (m *endpointManager) OnUpdate(msg interface{}) {
+	switch msg := msg.(type) {
+	case *proto.WorkloadEndpointUpdate:
+		log.WithField("workloadEndpointId", msg.Id).Info("Processing WorkloadEndpointUpdate")
+		m.pendingWlEpUpdates[*msg.Id] = msg.Endpoint
+	case *proto.WorkloadEndpointRemove:
+		log.WithField("workloadEndpointId", msg.Id).Info("Processing WorkloadEndpointRemove")
+		m.pendingWlEpUpdates[*msg.Id] = nil
+	case *proto.IPSetDeltaUpdate:
```

Comment:
Thanks. I hadn't seen a case (in testing) where updates came through as a IPSetUpdate instead of a IPSetDeltaUpdate, so I thought it was safe to have endpoint manager just react to the Deltas. If there are cases where IPSetUpdate might be generated instead, then we definitely need to handle this here. Added.